### PR TITLE
Add UI to select books from VitalSource

### PR DESCRIPTION
**Pull request by robertknight**
Modify the "Select book from VitalSource" option in the file picker app to show a book and chapter selection UI instead of submitting a hardcoded book. The data shown in the book and chapter selection UI is
currently hardcoded data, derived from the VitalSource Products API [1],
for the two books which we know are available.

 - Add a loading state (`isLoading`) to the `Table` component to indicate that data is currently being fetched. This is used by the `BookList` and `ChapterList` components added in this PR.
 - Add `BookPicker`, `BookList` and `ChapterList` components to support selecting a book from VitalSource and then a chapter within that book
 - Modify `ContentSelector` component to show the `BookPicker` when the "Select book from VitalSource" option is selected, instead of submitting a hardcoded book ID and chapter.
 - Add dummy book and chapter data from the VitalSource API in `vitalsource-sample-data.js`. This will in future be replaced by calls from the frontend to the backend which will in turn make a request to the VitalSource Products API.

[1] https://developer.vitalsource.com/hc/en-us/articles/360010967153-GET-v4-products-vbid-Title-TOC-Metadata

Part of https://github.com/hypothesis/product-backlog/issues/1200.

---

nb. In order to test these flows, you will need to go to http://localhost:8001/flags and turn on the "VitalSource" option locally.

When clicking the "Select book from VitalSource" button in the file picker, instead of selecting a hard-coded book, a list of "relevant" books is now fetched and shown. This list is currently hard-coded JSON data. In future this will be fetched from the backend. We may also introduce alternate flows where the user enters a search query or enters a book identifier or link:

Book picker 1

When a book is selected, the cover image is shown:

Book picker 2

When a book has been selected, the chapter list for that book is displayed. The chapter list data is also hard-coded JSON. In future this will be fetched from the backend.

Chapter picker

After the user has chosen a book and chapter to use, they are shown group set configuration. This screen is unchanged:

Book chosen

After the user clicks "Continue" in this screen, the `book_id` and `cfi` query params in the LTI launch URL will reflect the chosen book:

book_id query param
cfi query param

**TODO:**

- [x] Add screenshots to the PR description
- [x] Finish adding tests for new components
- [x] Fix issue where cover image can overflow container in the `BookList` component
- [x] Make the loading state look nicer when the `BookList` and `ChapterList` data is being loaded

### CODE REVIEW

File: lms/static/scripts/frontend_apps/components/BookList.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        columns={columns}
+        items={books}
+        onSelectItem={onSelectBook}
+        onUseItem={onUseBook}
+        renderItem={book => <td>{book.title}</td>}
+        selectedItem={selectedBook}
+      />
+      {selectedBook && (
+        <div className="BookList__cover-container">
+          <img
```
Comment (LMS007) [inline lms/static/scripts/frontend_apps/components/BookList.js:52]: Its a little distracting when you go from nothing selected, to then selecting a book from the list, then the list's width shrinks to make room for the image and potentially the heigh increases if there are not too many books in the list because the image may be taller. (as is the case with only 2 books)

The other thing is the image is centered vertically, but it might look better if it was fixed to the top. I mocked up some extra books to see how it would look.

Screen Shot 2021-05-27 at 3 15 46 PM


The other thing I noticed was that the overlfow:auto was not working. I looked at the css and I couldn't figure out why so that may be an issue with my mock data or some other css rule.

File: lms/static/scripts/frontend_apps/utils/api.js [hunk 2]
```diff
@@ -90,6 +94,36 @@ async function apiCall({ path, authToken, data }) {
...
+}
+
+/**
+ * Fetch a list of available ebooks to use in assignments.
+ *
+ * @param {string} authToken
+ * @param {number} [fetchDelay] - Dummy delay to simulate slow third-party
+ * @return {Promise<Book[]>}
+ */
+export async function fetchBooks(authToken, fetchDelay = 500) {
```
Comment (LMS007) [inline lms/static/scripts/frontend_apps/utils/api.js:109]: This is totally the kitchen drawer folder `utils`, and may need some TLC later. 

I realize these are placeholder, but when we replace them with the real calls, it may make sense to move them to their own file like the google api calls, grader-service, etc.

**Review by LMS007 — COMMENTED** [review]
Rob, these changes, while extensive, are very readable--Thank you. Most of my feedback is related to UX, the dialog sizing and the book image. I think there are some options that could improve things a bit, but I realize this is also just a first pass. At the very lest, we may want to use the old spinner and just hide the loading state which shows the squished dialog w/o content momentarily, or we could set the width of the dialog to a fixed size.

I also could not get the book to load, but I think thats my setup. In any case, I got a 422 error when picking a chapter.

File: lms/static/scripts/frontend_apps/utils/api.js (hunk 2 above)
Comment (robertknight) [reply lms/static/scripts/frontend_apps/utils/api.js:109]: > This is totally the kitchen drawer folder utils, and may need some TLC later.

Yes. I think it may be helpful to create a service layer in the LMS frontend and create a stateful service for API calls. That will enable us to avoid passing around the `authToken` to every function that makes API calls.

File: lms/static/scripts/frontend_apps/components/BookPicker.js [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+  const confirmChapter = useCallback(
+    chapter => {
+      onSelectBook(/** @type {Book} */ (book), chapter);
+    },
+    [book, onSelectBook]
+  );
+
+  useEffect(() => {
+    if (step === 'select-book' && !bookList) {
+      fetchBooks(authToken).then(setBookList).catch(setError);
```
[the remark this answers is not in the document]
Comment (robertknight) [reply lms/static/scripts/frontend_apps/components/BookPicker.js:56]: That's a good question. The assumption is that in future the request to fetch books will be parametrized to only show books that are associated with the current course, so even if the book store has millions of books, the number here shouldn't be large. There might be some kind of alternate flow for larger repositories where the user enters a search query and a list of results (up to say 100) is returned.

File: lms/static/scripts/frontend_apps/components/BookList.js (hunk 1 above)
Comment (robertknight) [reply lms/static/scripts/frontend_apps/components/BookList.js:48]: I've revised the UI here to fix the overflow issue. The book cover now has a fixed height of 180px and the width adapts depending on the book's aspect ratio, with a minimum width for the cover container of 180px. This means that for the majority of books that are taller than they are wide, the size of the book cover container will be fixed.

File: lms/static/scripts/frontend_apps/utils/api.js [hunk 4]
```diff
@@ -90,6 +94,36 @@ async function apiCall({ path, authToken, data }) {
...
 }
 
-// Separate export from declaration to work around
-// https://github.com/robertknight/babel-plugin-mockable-imports/issues/9
-export { apiCall };
+/** @param {number} ms */
+function delay(ms) {
+  return new Promise(resolve => setTimeout(resolve, ms));
+}
+
```
Comment (lyzadanger) [inline lms/static/scripts/frontend_apps/utils/api.js:101]: Could you add a comment to this function describing what it's for and maybe indicate that it's temporary?

File: lms/static/scripts/frontend_apps/utils/api.js [hunk 5]
```diff
@@ -90,6 +94,36 @@ async function apiCall({ path, authToken, data }) {
...
+ * @param {string} authToken
+ * @param {string} bookId
+ * @param {number} [fetchDelay] - Dummy delay to simulate slow third-party
+ * @return {Promise<Chapter[]>}
+ */
+export async function fetchChapters(authToken, bookId, fetchDelay = 500) {
+  await delay(fetchDelay);
+  if (!chapterData[bookId]) {
+    throw new ApiError(404, { message: 'Book not found' });
+  }
```
Comment (lyzadanger) [inline lms/static/scripts/frontend_apps/utils/api.js:127]: Is this branch a stub for a "real" error handler or to exercise a test path? i.e. is this an expected possible path in the current implementation?

File: lms/static/scripts/frontend_apps/utils/api.js (hunk 2 above)
Comment (lyzadanger) [reply lms/static/scripts/frontend_apps/utils/api.js:109]: I wonder if it might be worth extracting these in-flux VitalSource-related API mocks (as it were) into a separate module, even now? Their presence here in their current prototype states feels at odds with the production-class code in this module. I don't feel strongly about this, however.

File: lms/static/scripts/frontend_apps/utils/vitalsource-sample-data.js [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+      cfi: '/6/10[;vnd.vst.idref=ded]',
+      page: 'v',
+    },
+    {
+      title: 'Contents',
+      cfi: '/6/12[;vnd.vst.idref=toc]',
+      page: 'vii',
+    },
+    {
+      title: 'Forward',
```
Comment (lyzadanger) [inline lms/static/scripts/frontend_apps/utils/vitalsource-sample-data.js:290]: Super pedantic, but, hey, it totally caught my eye when tootling around the interface: this should be _Foreword_. It was likely misspelled in your source data, though?

File: lms/static/styles/components/_BookList.scss [hunk 7]
```diff
@@ -0,0 +1,31 @@
+// Height of cover images, chosen to work when the file picker app is shown
```
Comment (lyzadanger) [inline lms/static/styles/components/_BookList.scss:1]: I'm not diving in on this styling in this iteration...it gets the job done fine for now, and, with the exception of the "jumpiness" of the book-cover container, is OK to land as-is and we'll tune it later.

File: lms/static/styles/components/_Table.scss [hunk 8]
```diff
@@ -21,6 +21,15 @@ $table-item-height: 40px;
...
 
+.Table__spinner {
+  display: flex;
+  flex-direction: row;
+  justify-content: center;
+
+  padding-top: 10px;
+  padding-bottom: 10px;
+}
+
```
Comment (lyzadanger) [inline lms/static/styles/components/_Table.scss:32]: I'm thinking that standardizing the Spinner component/pattern might be useful in the near term in the shared package?

File: lms/static/scripts/frontend_apps/components/test/BookPicker-test.js [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+
+  const selectBook = wrapper => {
+    const bookList = wrapper.find('BookList');
+    const book = bookList.prop('books')[0];
+    act(() => {
+      bookList.prop('onSelectBook')(book);
+    });
+    wrapper.update();
+    return book;
+  };
```
Comment (lyzadanger) [inline lms/static/scripts/frontend_apps/components/test/BookPicker-test.js:45]: In the manner that `clickSelectButton` below is crystal-clear naming, would this more accurately be `selectFirstBook`? It might help a dev scan more quickly here.

File: lms/static/scripts/frontend_apps/components/test/BookPicker-test.js [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+
+  const selectChapter = wrapper => {
+    const chapterList = wrapper.find('ChapterList');
+    const chapter = chapterList.prop('chapters')[0];
+    act(() => {
+      chapterList.prop('onSelectChapter')(chapter);
+    });
+    wrapper.update();
+    return chapter;
+  };
```
Comment (lyzadanger) [inline lms/static/scripts/frontend_apps/components/test/BookPicker-test.js:55]: `selectFirstChapter`?

File: lms/static/scripts/frontend_apps/components/test/BookPicker-test.js [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+        .prop('onClick')();
+    });
+    wrapper.update();
+  };
+
+  it('fetches and displays book list when picker is opened', async () => {
+    const picker = renderBookPicker();
+
+    let bookList = picker.find('BookList');
+    assert.deepEqual(bookList.prop('books'), []);
```
Comment (lyzadanger) [inline lms/static/scripts/frontend_apps/components/test/BookPicker-test.js:70]: This is specifically testing that the book list is empty when the picker is opened. Is it worthy of its own test, or commenting, or anything?

File: lms/static/scripts/frontend_apps/components/test/BookPicker-test.js [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+    let chapterList = picker.find('ChapterList');
+    assert.isTrue(chapterList.exists());
+    assert.equal(chapterList.prop('isLoading'), true);
+    assert.calledWith(fakeAPI.fetchChapters, 'dummy-auth-token');
+
+    await waitForElement(picker, 'ChapterList[isLoading=false]');
+
+    chapterList = picker.find('ChapterList');
+    assert.equal(chapterList.prop('chapters'), bookData.chapterData[book.id]);
+  });
```
Comment (lyzadanger) [inline lms/static/scripts/frontend_apps/components/test/BookPicker-test.js:97]: This set of tests also tests the selection callback (i.e. what happens when a chapter is selected) in addition to what its test description captures. i.e perhaps there is more than one test here?

File: lms/static/scripts/frontend_apps/components/test/BookList-test.js [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+import BookList from '../BookList';
+
+describe('BookList', () => {
+  const bookData = [
+    {
+      id: 'test-book',
+      title: 'Test book',
+      cover_image: 'https://example.com/test.jpg',
+    },
+  ];
```
Comment (lyzadanger) [inline lms/static/scripts/frontend_apps/components/test/BookList-test.js:13]: One day we might want to talk about fixtures for our apps...it's something I've had back-of-mind for a while.

File: lms/static/scripts/frontend_apps/components/test/BookList-test.js [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+    const bookList = renderBookList();
+    assert.equal(bookList.find('tbody > tr').length, bookData.length);
+    assert.include(bookList.html(), '<td>Test book</td>');
+  });
+
+  [true, false].forEach(isLoading => {
+    it('shows loading indicator if books are being fetched', () => {
+      const bookList = renderBookList({ isLoading });
+      assert.equal(bookList.find('Table').prop('isLoading'), isLoading);
+    });
```
Comment (lyzadanger) [inline lms/static/scripts/frontend_apps/components/test/BookList-test.js:37]: Ah! The test description here had me ready to comment "Is this testing `Table` functionality?" But then I realized the test itself is appropriate for `Booklist` (making sure that the `isLoading` prop is set correctly). I wonder if adjusting the test description would help, as the test is not about the loading indicator display (which is a responsibility of `Table`), though that is the manifest result.

(There is a similar test for `ChapterList`)

File: lms/static/scripts/frontend_apps/components/test/ContentSelector-test.js [hunk 15]
```diff
@@ -365,10 +365,27 @@ describe('ContentSelector', () => {
       button.props().onClick();
     });
 
+    const bookPicker = wrapper.find('BookPicker');
+    assert.equal(bookPicker.prop('authToken'), fakeConfig.api.authToken);
+    assert.isTrue(wrapper.exists('BookPicker'));
+  });
+
+  it('submits selecting a book from VitalSource', () => {
```
Comment (lyzadanger) [inline lms/static/scripts/frontend_apps/components/test/ContentSelector-test.js:373]: This test description is a little hard to parse.

File: lms/static/scripts/frontend_apps/utils/test/api-test.js [hunk 16]
```diff
@@ -125,4 +125,35 @@ describe('api', () => {
...
+
+  describe('fetchChapters', () => {
+    it('returns chapter list', async () => {
+      const data = await fetchChapters(
+        'auth-token',
+        'BOOKSHELF-TUTORIAL',
+        1 /* delay */
+      );
+      assert.isTrue(Array.isArray(data));
+      assert.equal(data.length, 45);
```
Comment (lyzadanger) [inline lms/static/scripts/frontend_apps/utils/test/api-test.js:145]: 45 is a little magic, but I'll allow it at this stage in the game ;)

**Review by lyzadanger — APPROVED** [review]
This is a big PR, but I’ve done my best to be attentive to its various bits. The set of components is well-organized, easy enough to read through, and tests are clear.

To set context, I’ve reviewed the code and functionality through a “stable prototype” filter. The expectations here are that code is stable, tested, does what it’s supposed to, and the UI is usable, if not pretty. I’ve made certain allowances for slight inelegances and areas I can see will be addressed in future iterations. Nothing I found is a must-fix to land within this context.

On the UI theme, the one thing that did jump out at me in this current implementation (i.e. it would be lovely to fix before landing, if not too taxing) is some jumpiness in the book-selection step. In my testing, the list of books would populate the table initially, with the title spanning full width. When clicking on a book, the book-cover image suddenly appears, takes up part of the table width, and changes the width of the title column abruptly. Would it be possible to give the cover-image container dimensions such that it doesn’t appear and disappear?

Other UI details like padding, alignment, other jumpiness, etc. were noted but don’t need to be addressed in this changeset.

Kyle also noticed in his review that the `api` module is feeling a little cluttered, and I’m inclined to agree. I added a suggestion to perhaps extract the stubby/prototype-y book- and chapter-fetching logic into a separate module, though this is only a suggestion and I’m not dead set on it.

It feels like the LMS front-end app is starting to get to another growth stage in its development, and there are a few places where it’s straining at the seams: state management, API encapsulation, multi-step form-flow management. None of these require immediate intervention, but perhaps it’s coming time to have a group think about the app evolution/strategy?

File: lms/static/scripts/frontend_apps/components/BookPicker.js [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+            if (step === 'select-book' && book) {
+              confirmBook(book);
+            } else if (step === 'select-chapter' && chapter) {
+              confirmChapter(chapter);
+            }
+          }}
+          variant="primary"
+        >
+          {step === 'select-book' ? 'Select book' : 'Select chapter'}
+        </LabeledButton>,
```
[the remark this answers is not in the document]
Comment (robertknight) [reply lms/static/scripts/frontend_apps/components/BookPicker.js:90]: The `&& book` and `&& chapter` logic in the conditionals is strictly redundant. TS can't make the connection though that `disabled={true}` implies that `onClick` will not be called. I'll rewrite this to make this assumption more obvious and avoid the redundant runtime check.

File: lms/static/scripts/frontend_apps/components/BookPicker.js [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+  const [bookList, setBookList] = useState(/** @type {Book[]|null} */ (null));
+  const [chapterList, setChapterList] = useState(
+    /** @type {Chapter[]|null} */ (null)
+  );
+  const [book, setBook] = useState(/** @type {Book|null} */ (null));
+  const [chapter, setChapter] = useState(/** @type {Chapter|null} */ (null));
+  const [step, setStep] = useState(
+    /** @type {'select-book'|'select-chapter'} */ ('select-book')
+  );
+  const [error, setError] = useState(/** @type {Error|null} */ (null));
```
[the remark this answers is not in the document]
Comment (robertknight) [reply lms/static/scripts/frontend_apps/components/BookPicker.js:40]: I don't think its a problem at present. The thing I would pay more attention to rather than the number of state values is whether there is redundancy (eg. one state value always implying another) or whether the state structure makes it possible to represent states that should be impossible. In that case some refactoring is called for.

You could potentially push some of this state down into the `BookList` and `ChapterList` components and let them "own" the book / chapter list. If so however you then have an issue that you'd want to avoid re-fetching the list if we added a "Back" button to the dialog in future (that is, go back from the chapter selection screen to the book selection screen).

File: lms/static/scripts/frontend_apps/components/ChapterList.js [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+
+/**
+ * @typedef ChapterListProps
+ * @prop {Chapter[]} chapters - List of available chapters
+ * @prop {boolean} [isLoading] - Whether to show a loading indicator
+ * @prop {Chapter|null} selectedChapter - The chapter within `chapters` which is currently selected
+ * @prop {(c: Chapter) => void} onSelectChapter - Callback invoked when the user selects a chapter
+ * @prop {(c: Chapter) => void} onUseChapter -
+ *   Callback invoked when user confirms they want to use the selected chapter for
+ *   an assignment.
```
[the remark this answers is not in the document]
Comment (robertknight) [reply lms/static/scripts/frontend_apps/components/ChapterList.js:17]: I'm not super happy with the terminology here. "select" means that the user has selected a row in the table with a single click or tap or keyboard and "use" means that the user has pressed Enter with a row selected or double-clicked a row to indicate that they want to use this particular itme.

File: lms/static/scripts/frontend_apps/utils/api.js (hunk 4 above)
Comment (robertknight) [reply lms/static/scripts/frontend_apps/utils/api.js:101]: `delay` is a helper that we reproduce quite widely in tests to wait for a fixed amount of time. I'll add a comment to `fetchBooks` and `fetchChapters` to be more explicit that they are fakes/stubs.

File: lms/static/scripts/frontend_apps/utils/vitalsource-sample-data.js (hunk 6 above)
Comment (robertknight) [reply lms/static/scripts/frontend_apps/utils/vitalsource-sample-data.js:290]: Yes. I took the source data verbatim. I spotted various other apparent errors in the data as well, such as duplicated chapters. This might be something we have to somehow deal with on the server-side by filtering it before serving it to the client.

File: lms/static/styles/components/_Table.scss (hunk 8 above)
Comment (robertknight) [reply lms/static/styles/components/_Table.scss:32]: Yes, I think so.

File: lms/static/scripts/frontend_apps/components/test/BookPicker-test.js (hunk 9 above)
Comment (robertknight) [reply lms/static/scripts/frontend_apps/components/test/BookPicker-test.js:45]: Hmm... it might be useful to have tests that select other books in future. I think I can achieve the desired result by replacing the hard-coded `0` with an `index` which is a parameter with a default value of `0`.

File: lms/static/scripts/frontend_apps/components/test/BookPicker-test.js (hunk 12 above)
Comment (robertknight) [reply lms/static/scripts/frontend_apps/components/test/BookPicker-test.js:97]: This test does check what happens when a book is selected, but that's necessary to get the chapter list to appear in a way that realistically simulates the user. It doesn't check what happens when a chapter is selected though. That's in the next test.

File: lms/static/scripts/frontend_apps/components/test/ContentSelector-test.js (hunk 15 above)
Comment (robertknight) [reply lms/static/scripts/frontend_apps/components/test/ContentSelector-test.js:373]: Agh. Indeed. Will rewrite.

File: lms/static/scripts/frontend_apps/components/test/BookList-test.js (hunk 14 above)
Comment (robertknight) [reply lms/static/scripts/frontend_apps/components/test/BookList-test.js:37]: Ah OK. I'll rephrase this.

**Comment by robertknight** [comment]
I pushed changes to improve various test descriptions and comments in response to the comments above. I also made a change to always show the book cover container element, whether a book is selected or not. This makes the layout more consistent between the book selected/book not selected states.

> Kyle also noticed in his review that the api module is feeling a little cluttered, and I’m inclined to agree. I added a suggestion to perhaps extract the stubby/prototype-y book- and chapter-fetching logic into a separate module, though this is only a suggestion and I’m not dead set on it.

Logically I think it would make sense to separate this into two layers: one that is a low-level API client and one that is a higher level client for specific needs. A separate refactoring dimension might be to add a service layer and then have a stateful API service. This would allow the API service to deal with authentication without needing the caller to pass it in each time.

I'd like to leave this refactoring until later though as it might be useful to create API wrappers for some of the other calls that the client makes.

> None of these require immediate intervention, but perhaps it’s coming time to have a group think about the app evolution/strategy?

Yes. My rough thinking is that we could use the Hypothesis sidebar app as an architecture template and add the useful parts of that to other applications at they become complex enough to benefit from it. For example: a service layer, a dedicated model/store layer.

File: lms/static/scripts/frontend_apps/components/BookList.js (hunk 1 above)
Comment (robertknight) [reply lms/static/scripts/frontend_apps/components/BookList.js:53]: I've revised this so that the book cover container is always present, but is empty when no book is selected.

File: lms/static/scripts/frontend_apps/components/BookPicker.js [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+        .then(setChapterList)
+        .catch(setError);
+    }
+  }, [authToken, book, bookList, step, chapterList]);
+
+  const canSubmit =
+    (step === 'select-book' && book) || (step === 'select-chapter' && chapter);
+
+  return (
+    <Dialog
```
[the remark this answers is not in the document]
Comment (robertknight) [reply lms/static/scripts/frontend_apps/components/BookPicker.js:77]: > Perhaps a simple css fix could work by forcing the width to min(700px, 90vw); to be continued ....

Setting a minimum size on the dialog was the solution I had in a prototype of this PR at one point. For the moment I've left this as a follow-up task.

File: lms/static/scripts/frontend_apps/components/test/BookPicker-test.js (hunk 11 above)
Comment (robertknight) [reply lms/static/scripts/frontend_apps/components/test/BookPicker-test.js:71]: I've added some comments here to highlight the different phases of this test.

File: lms/static/scripts/frontend_apps/utils/api.js (hunk 2 above)
Comment (robertknight) [reply lms/static/scripts/frontend_apps/utils/api.js:112]: I wrote about this in a separate comment, but I think there will be some refactoring of other API-calling code that we might want to do at the same time for consistency. For example, we might require all calls to the API to go through wrapper functions. I've left this as a follow-up exercise for the moment.
